### PR TITLE
[SYCL][Driver] Enhance -Xarch_device and -Xarch_host for sycl

### PR DESCRIPTION
**Pull request by jinge90**
Add simple support to enhance -Xarch_device and -Xarch_host in sycl offloading which can support multiple arguments in single '-Xarch_*' and '-mllvm <opt>' is supported as well.

### CODE REVIEW

**Comment by jinge90** [comment]
Hi, @mdtoguchi and @AaronBallman 
As we discussed before, this draft patch implements "--target-options=" and "--host-options=", could you help review whether the work is in correct direction?
Thanks very much.

**Comment by mdtoguchi** [comment]
I encountered other options:  `-Xarch_host <arg>` and `-Xarch_device <arg>` which can be used to push options to either the host or device compilation.  Quick checking shows this works for various options, but does not work for any options that take arguments (like `-mllvm <opt>`).  Would it make more sense to leverage that instead of introducing new options?

**Comment by jinge90** [comment]
> -mllvm <opt>

Hi, @mdtoguchi 
Nice finding! I think -Xarch_device and -Xarch_host have almost same behavior as --target-options and --host-options except flag such as '-mllvm <opt>' and if users want to pass a series of flags, they need to type -Xarch_device or -Xarch_host multiple times.
Do you think we need to enhance it or not?
Thanks very much.

File: clang/lib/Driver/Compilation.cpp [hunk 1]
```diff
@@ -60,13 +60,64 @@ Compilation::~Compilation() {
       delete Arg.second;
 }
 
+static void HandleXarchArgs(DerivedArgList *OffloadArgList, const Driver &D) {
+  if (!OffloadArgList || (!OffloadArgList->hasArg(options::OPT_Xarch_device) &&
+                          !OffloadArgList->hasArg(options::OPT_Xarch_host)))
+    return;
+
+  bool IsDevice = OffloadArgList->hasArg(options::OPT_Xarch_device);
```
Comment (mdtoguchi) [inline clang/lib/Driver/Compilation.cpp:68]: Does this handle the possibility of passing both `-Xarch_host` and `-Xarch_device` on the same command line?

Comment (jinge90) [reply clang/lib/Driver/Compilation.cpp]: Hi, @mdtoguchi 
Thanks for pointing out this, I have updated patch to handle combination use of -Xarch_device and -Xarch_host and also added lit tests.
Thanks very much.

File: clang/test/Driver/sycl-xarch.cpp [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+// RUN:   | FileCheck %s -check-prefix=SYCL_XARCH_HOST_OPTIONS2
+// RUN: %clangxx -fsycl %s -Xarch_host "-fsanitize=address -DXARCH_HOST_TEST -mllvm -enable-merge-functions" -### 2>&1 \
+// RUN:   | FileCheck %s -check-prefix=SYCL_XARCH_HOST_OPTIONS3
+// SYCL_XARCH_HOST_OPTIONS1: clang{{.*}} "-fsycl-is-host"
+// SYCL_XARCH_HOST_OPTIONS1-SAME: -fsanitize=address
+// SYCL_XARCH_HOST_OPTIONS1-SAME: -fsanitize-address-use-after-scope
+// SYCL_XARCH_HOST_OPTIONS2: clang{{.*}} "-fsycl-is-host"
+// SYCL_XARCH_HOST_OPTIONS2-SAME: XARCH_HOST_TEST
+// SYCL_XARCH_HOST_OPTIONS3: clang{{.*}} "-fsycl-is-host"
+// SYCL_XARCH_HOST_OPTIONS3-SAME: "-mllvm" "-enable-merge-functions"
```
Comment (mdtoguchi) [inline clang/test/Driver/sycl-xarch.cpp:76]: Could you add a test that verifies `-Xarch_host "-mllvm -enable-merge-functions" -Xarch_host -DFOO` behaviors (i.e. passing multiple -Xarch_host or -Xarch_device options on the same command line)

Comment (jinge90) [reply clang/test/Driver/sycl-xarch.cpp:76]: Hi, @mdtoguchi 
Done. Thanks very much.

**Review by mdtoguchi — APPROVED** [review]
This looks OK to me.  I did notice that currently this implementation does not apply to Windows.  Moving forward, we should add Windows support.

**Comment by jinge90** [comment]
Hi, @intel/llvm-gatekeepers 
Could you help merge this patch?
Thanks very much.

**Pull request merged**